Source/WebCore/dom/Element.cpp

```diff
@@ -5531,4 +5532,25 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
     return target;
 }
 
+bool Element::hasCustomState(const AtomString* state) const
+{
+    if (!document().settings().customStateSetEnabled())
+        return false;
+
+    if (hasRareData()) {
+        CustomStateSet* customStates = elementRareData()->customStateSet();
+        return customStates ? customStates->has(*state) : false;
+    }
+
+    return false;
+}
+
+CustomStateSet* Element::ensureCustomStateSet()
```

Comment:
Why does this return a raw pointer? It doesn't seem it can return null. If so, this should return a `CustomStateSet&`.